### DISCUSSION TITLE
gcc/11 Compilation Issue

### DISCUSSION THREAD

**Discussion by laocoon91** [SPECFEM/specfem3d#1584] [SPECFEM3D General]
Hello,

I am attempting to recompile specfem on a machine that recently updated to gcc/11 from gcc/8. I've been getting inundated in the following errors before the compilation exits out:

> src/shared/parallel.f90:334:17:
>
>   334 |   call MPI_BCAST(buffer,1,MPI_INTEGER,0,my_local_mpi_comm_world,ier)
>       |                 1
> ......
>   352 |   call MPI_BCAST(buffer,1,MPI_LOGICAL,0,my_local_mpi_comm_world,ier)
>       |                 2
> Error: Type mismatch between actual argument at (1) and actual argument at (2) (INTEGER(4)/LOGICAL(4)).

Looking around online, I've seen suggestions to append FCFLAGS="-w -fallow-argument-mismatch -O2" or removing -Werror when configuring, but this has not resolved the error for me. Has anyone else encountered this issue? For reference, my configuration command is:

>  ./configure FC=gfortran CC=gcc MPIFC=ftn --with-mpi MPI_INC="${CRAY_MPICH2_DIR}/include" FCFLAGS="-w -fallow-argument-mismatch -O2"

It is a cray machine, but I am using a gnu programming environment. 

Thank you,

Ian

**Comment by danielpeter** [comment]
thanks for bringing this up! 

yes, from newer GCC 10 compilers onward, they introduced this stricter argument checking in single files which leads to an issue with the MPI calls. until GCC and MPI folks figure out how to fix this in upcoming versions, one obvious workaround is to use a GCC compiler version < 10. if you have such an older module still available on your Cray system, like `gcc/9.3.0` on my Cray XC40 system, you could just type for example:
```
module switch gcc gcc/9.3.0 
```
and re-configure/compile as usual without special flags.

for newer GCC versions, there is either the `-std=legacy` or the  `-fallow-argument-mismatch` flag which can be added to convert such error messages into warnings. this would work on my Cray system:
```
./configure FC=gfortran CC=gcc MPIFC=ftn --with-mpi MPI_INC="${CRAY_MPICH2_DIR}/include" FLAGS_CHECK="-O3 -fallow-argument-mismatch -Wall -Wno-conversion -Wno-maybe-uninitialized"
```
or with the `-std=legacy` flag:
```
./configure FC=gfortran CC=gcc MPIFC=ftn --with-mpi MPI_INC="${CRAY_MPICH2_DIR}/include" FLAGS_CHECK="-O3 -std=legacy -Wall -Wno-conversion -Wno-maybe-uninitialized"
```
give it a try and let me know if this solves your issue. 

i hope that GCC and MPI will work this out in future releases, otherwise, we have to put every MPI wrapper call into its own file, which would be a bit annoying...

**Reply by laocoon91** [reply]
The -fallow-argument-mismatch flag combination you recommended successfully converted the errors to warnings, and everything looks to be working ok now. Thank you for the help (and the quick reply)!
